### PR TITLE
Fix cloning fails in some cases for organizations

### PR DESCRIPTION
**Pull request by sharat87**
Cloning currently fails in cases like the following:

- Application with no pages
- Pages with no actions
- Pages with more than one action

This PR fixes cases like these as well as adds tests for them and a few more such scenarios.

Additionally, the implementation of adding a page to an application has been changed to avoid race-conditions that can occur when two pages are being added to an application in separate reactive flows.

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java [hunk 1]
```diff
@@ -87,34 +98,37 @@ public ApplicationPageServiceImpl(ApplicationService applicationService,
...
-    public Mono<Application> addPageToApplication(Mono<Application> applicationMono, Page page, Boolean isDefault) {
-        return applicationMono
-                .map(application -> {
-                    List<ApplicationPage> applicationPages = application.getPages();
-                    if (applicationPages == null) {
-                        applicationPages = new ArrayList<>();
+    @Override
+    public Mono<UpdateResult> addPageToApplication(Application application, Page page, Boolean isDefault) {
+        final ApplicationPage applicationPage = new ApplicationPage(page.getId(), isDefault);
+        return mongoTemplate
```
Comment (mohanarpit) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java:121]: Is it possible for us to move `mongoTemplate` query to the `CustomApplicationRepositoryImpl` ? It'll ensure that all DB queries for a given collection are located in a single class.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java [hunk 2]
```diff
@@ -87,34 +98,37 @@ public ApplicationPageServiceImpl(ApplicationService applicationService,
...
+                .updateFirst(
+                        Query.query(Criteria.where("_id").is(application.getId())),
+                        new Update().addToSet("pages", applicationPage),
+                        Application.class
+                )
+                .doOnSuccess(result -> {
+                    if (result.getModifiedCount() != 1) {
```
Comment (mohanarpit) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java:128]: Minor: Should this check be `< 1 ` or is it actually ` != 1 `?

File: app/server/appsmith-server/src/test/java/com/appsmith/server/solutions/ExamplesOrganizationClonerTests.java [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+                            "action1",
+                            "action2",
+                            "action3",
+                            "action4"
+                    );
+                })
+                .verifyComplete();
+    }
+
+    private <InType, OutType> List<OutType> map(List<InType> list, Function<InType, OutType> fn) {
```
Comment (mohanarpit) [inline app/server/appsmith-server/src/test/java/com/appsmith/server/solutions/ExamplesOrganizationClonerTests.java:545]: Can you help me understand what this function is for?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java (hunk 1 above)
Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java:121]: Yeah good point. Will do.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java (hunk 2 above)
Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java:128]: Hm, if it is `> 1` that means there's two applications with the same ID, which is also an error. So, any value but `1` should be treated as an error right?

File: app/server/appsmith-server/src/test/java/com/appsmith/server/solutions/ExamplesOrganizationClonerTests.java (hunk 3 above)
Comment (sharat87) [reply app/server/appsmith-server/src/test/java/com/appsmith/server/solutions/ExamplesOrganizationClonerTests.java:545]: Just a handy test utility function to call a function over each object in `list`, collect the return values in a list and return that list.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java (hunk 1 above)
Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java]: Done

File: app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/CustomApplicationRepositoryImpl.java [hunk 4]
```diff
@@ -64,4 +70,14 @@ protected Criteria getIdCriteria(Object id) {
         return queryAll(List.of(orgIdsCriteria), permission);
     }
 
+    @Override
+    public Mono<UpdateResult> addPageToApplication(Application application, Page page, boolean isDefault) {
+        final ApplicationPage applicationPage = new ApplicationPage(page.getId(), isDefault);
+        return mongoOperations.updateFirst(
+                Query.query(Criteria.where("_id").is(application.getId())),
```
Comment (mohanarpit) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/CustomApplicationRepositoryImpl.java:77]:
```suggestion
                Criteria idCriteria = getIdCriteria(application.getId());
                Query.query(idCriteria),
```

**Pull request merged**